R/run-sims.R

```diff
@@ -0,0 +1,383 @@
+      data <- bbr::nm_data(mod)
+    })
+  }
+
+  data_cols <- names(data)
+
+  if (!join_col %in% data_cols) {
+    stop("`join_col` (", join_col, ") not found in data columns.")
+  }
+
+  if (length(intersect(RUN_SIMS_COLS, data_cols))) {
+    stop("Some data names collide with bbr-bayes-run-sims.tab names: ",
+         paste(intersect(RUN_SIMS_COLS, data_cols), collapse = ", "))
+  }
+
+  if (!"EVID" %in% data_cols) {
+    data$EVID <- 0
+  }
+
+  tab_files <- chain_paths_impl(mod,
+                                name = "bbr-bayes-run-sims",
+                                extension = "tab",
+                                check_exists = "all_or_none")
+  if (!length(tab_files)) {
+    stop("Chain models do not have bbr-bayes-run-sims.tab files\n",
+         "This is unexpected unless you called `submit_model()`\n",
+         "with `.run_sims_col = NULL`.")
+  }
+
+  tab <- purrr::map(tab_files, fread_chain_file) %>%
+    dplyr::bind_rows() %>%
+    dplyr::select(-"DV")
+
+  if (!join_col %in% names(tab)) {
+    stop("`join_col` (", join_col, ") is not in bbr-bayes-run-sims.tab\n",
+         "See `?bbr.bayes::nmbayes_submit_model`.")
+  }
+
+  tab_sum <- dplyr::group_by(tab, .data[[join_col]]) %>%
+    dplyr::summarise(dplyr::across(dplyr::everything(), .fns = point_fn))
+
+  dplyr::left_join(tab_sum, data, by = join_col)
+}
+
+#' Randomly select ext samples
+#'
+#' @param mod A `bbi_nmbayes_model` object.
+#' @param n_post Number of posterior samples to select.
+#' @param min_batch_size Split the samples into a list of data frames so that
+#'   each element has at least this number of samples.
+#'
+#' @return List of data frames with ext samples, split for sending into `future`
+#'   map functions.
+#' @noRd
+sample_exts <- function(mod, n_post, min_batch_size) {
+  # Note: Downstream mrgsolve wants a data frame in the same form as the ext
+  # data files. Read these directly rather than using as_draws_df() to avoid
+  # needing to do extra work to get back to this same place.
+  ext <- chain_paths_impl(mod, extension = "ext", check_exists = "all") %>%
+    purrr::map(fread_chain_file) %>%
+    dplyr::bind_rows(.id = "chain") %>%
+    dplyr::filter(.data$ITERATION > 0) %>%
+    dplyr::mutate(draw = dplyr::row_number())
+
+  if (nrow(ext) > n_post) {
+    ext <- ext[sort(sample(nrow(ext), n_post)), ]
+  }
+
+  nrows <- nrow(ext)
+  nsplits <- max(nrows %/% min_batch_size, 1)
+  purrr::map(parallel::splitIndices(nrows, nsplits), function(i) ext[i, ])
+}
+
+sim_epred <- function(mod_mrgsolve, exts, data, join_col, y_col, pbar) {
+  theta_cols <- grep("^THETA[0-9]+$", colnames(exts[[1]]))
+  mod_sim <- mrgsolve::data_set(mod_mrgsolve, data)
+  res <- future.apply::future_lapply(
+    exts,
+    function(ext) {
+      `%>%` <- magrittr::`%>%` # Make pipe available in other sessions.
+      purrr::map(seq_len(nrow(ext)), function(n) {
+        ext_row <- ext[n, ]
+        pbar(sprintf("EPRED (%d)", ext_row$draw))
```

Comment:
The draw is the the row number of the samples concatenated across chains (i.e. same as `posterior::draws_ids()`).  Here we have a random set of draws, in increasing order.

I initially liked the idea of displaying it in the progress bar text.  However, on second thought, I think it's just going to cause confusion to see the numbers increasing with random jumps.  It'd be better for this to use `n`.  Same for ipred progress bar below.

Reply:
> It'd be better for this to use `n`

Hmm, I guess that's confusing to because it's just the iteration number within a batch.  Just dropped the number part completely: b71e3dd